### DISCUSSION TITLE
How to experiment/debug without losing data?

### DISCUSSION THREAD

**Discussion by vtim** [johanmeijer/grott#291] [Q&A]
Hello @johanmeijer 👋

I found Grott while doing research because I had the idea to write something similar 🙃.
Looks like you’ve built quite a nice tool, congrats!

It’s not really necessary any more, but I’d still like to write a parser for the binary format in Elixir as a learning exercise.
I was wondering if/how you were able to test and develop without losing data? I’d like to keep sending all data to the Growatt servers until I’m certain I have a fully working local setup that I know can keep running.

Thanks in advance!
—Tim

**Comment by johanmeijer** [comment]
I start writing grott based on a sniffer technic. I routed the network traffic via my rpi (described it in the wiki) and sniffed it to start analysing the protocol (offcourse also found some info on internet of people who were doing this).

After that I created a proxy and used that. The advantage of the proxy is,  next that is easier to install,  that you also can block data (commands) comming from China.

I now have 2 test programs. One that analyse the data record (without sending data to growatt). This program is published in the examples directory (testoffset.py).
The other can fully emulate and validate the complete datastream. Using this together with grottserver (emulating the growatt website) can perform a full end2end test.
I haven't published this while it is not nicely structured and easy to use.
